rmf_traffic/test/unit/agv/test_Negotiator.cpp

```diff
@@ -204,59 +205,97 @@ SCENARIO("Test Plan Negotiation Between Two Participants")
+      negotiator_1.respond(
+        next_table->viewer(),
+        rmf_traffic::schedule::SimpleResponder::make(next_table));
 
-    CHECK(negotiation->ready());
+      CHECK(negotiation->ready());
 
-    next_table = negotiation->table(p2.id(), {p1.id()});
-    negotiator_2.respond(
-      next_table->viewer(),
-      rmf_traffic::schedule::SimpleResponder::make(next_table));
+      next_table = negotiation->table(p2.id(), {p1.id()});
+      negotiator_2.respond(
+        next_table->viewer(),
+        rmf_traffic::schedule::SimpleResponder::make(next_table));
 
-    CHECK(negotiation->complete());
+      CHECK(negotiation->complete());
 
-    auto proposals = negotiation->evaluate(
-      rmf_traffic::schedule::QuickestFinishEvaluator())->proposal();
-    REQUIRE(proposals.size() == 2);
+      auto proposals = negotiation->evaluate(
+        rmf_traffic::schedule::QuickestFinishEvaluator())->proposal();
+      REQUIRE(proposals.size() == 2);
 
-    const auto& proposal_1 = proposals.at(0);
-    const auto& proposal_2 = proposals.at(1);
-    for (const auto& r1 : proposal_1.itinerary)
-    {
-      for (const auto& r2 : proposal_2.itinerary)
+      const auto& proposal_1 = proposals.at(0);
+      const auto& proposal_2 = proposals.at(1);
+      for (const auto& r1 : proposal_1.itinerary)
       {
-        CHECK_FALSE(rmf_traffic::DetectConflict::between(
-            profile, r1->trajectory(),
-            profile, r2->trajectory()));
+        for (const auto& r2 : proposal_2.itinerary)
+        {
+          CHECK_FALSE(rmf_traffic::DetectConflict::between(
+              profile, r1->trajectory(),
+              profile, r2->trajectory()));
+        }
       }
+
+      //print_proposal(*proposals);
     }
 
-//    print_proposal(*proposals);
+    GIVEN("Negotiator #2 is a StubbornNegotiator")
+    {
+      rmf_traffic::schedule::StubbornNegotiator negotiator_2{ p2 };
+
+      auto next_table = negotiation->table(p1.id(), {});
+      negotiator_1.respond(
+        next_table->viewer(),
+        rmf_traffic::schedule::SimpleResponder::make(next_table));
+
+      REQUIRE(negotiation->table(p2.id(), {p1.id()}));
+
+      next_table = negotiation->table(p2.id(), {});
+      negotiator_2.respond(
+        next_table->viewer(),
+        rmf_traffic::schedule::SimpleResponder::make(next_table));
+
+      REQUIRE(negotiation->table(p1.id(), {p2.id()}));
+
+      next_table = negotiation->table(p1.id(), {p2.id()});
+      negotiator_1.respond(
+        next_table->viewer(),
+        rmf_traffic::schedule::SimpleResponder::make(next_table));
+
+      CHECK(negotiation->ready());
+
+      next_table = negotiation->table(p2.id(), {p1.id()});
+      // there should be a rejection of the alternative itinerary
+      negotiator_2.respond(
+        next_table->viewer(),
+        rmf_traffic::schedule::SimpleResponder::make(next_table));
+
+      // in this scenario, the negotiation never completes
+      CHECK_FALSE(negotiation->complete());
```

Comment:
These expectations are all good, but after this `CHECK_FALSE` you could continue the negotiation process to completion.

After the `negotiator_2` rejects table `[0]`, you can give table `[0]` back to `negotiator_1`, and it should be able to give a proposal that accommodates `negotiator_2`. At that point, giving table `[0, 1]` to `negotiator_2` should succeed, and then the negotiation should be complete.

Reply:
Note that what I described works because when `negotiator_2` rejects table `[0]` it also leaves a hint for `negotiator_1` about how `negotiator_1` can plan around it. Providing that hint is the key difference between "reject" and "forfeit".

Reply:
>  At that point, giving table [0, 1] to negotiator_2 should succeed, and then the negotiation should be complete.

(1) If I'm doing this right, adding the following after `CHECK_FALSE(negotiation->complete());` yields an incomplete solution despite it not going down the StubbornNegotiator::respond rejection path.
```
next_table = negotiation->table(p1.id(), {p2.id()});
negotiator_2.respond(
        next_table->viewer(),
        rmf_traffic::schedule::SimpleResponder::make(next_table)); //not rejected

CHECK(negotiation->complete()); //fails
```

(2) whereas the following succeeds. Correct me if I'm wrong, but it seems negotiator_2 has accepted the original p1's proposals.
```
next_table = negotiation->table(p1.id(), {});
negotiator_2.respond(
    next_table->viewer(),
    rmf_traffic::schedule::SimpleResponder::make(next_table));
CHECK(negotiation->complete()); //succeeds
```

This code crashes, and I suppose there's no table `[1, 0]` it's rejected.
```
next_table = negotiation->table(p2.id(), {p1.id()});
```

Let me know if anything's wrong with my assumptions. I'm putting (2) into the commit.

Reply:
What you should add after `CHECK_FALSE(negotiation->complete());` is the following:

```
next_table = negotiation->table(p1.id(), {});
negotiator_1.respond(
  next_table->viewer(),
  rmf_traffic::schedule::SimpleResponder::make(next_table));

// TODO: Test that negotiator_1 found a solution

next_table = negotiation->table(p2.id(), {p1.id()});
negotiator_2.respond(
  next_table->viewer(),
  rmf_traffic::schedule::SimpleResponder::make(next_table));

// TODO: Test that the negotiation is complete
```

You need to have `negotiator_1` respond again. In the code that you've pushed, you're having `negotiator_2` respond to a table that belongs to `negotiator_1`.

Reply:
This LOC `next_table = negotiation->table(p2.id(), {p1.id()});` returns a null pointer.

Reply:
That's problematic. Please push your latest version of the test with the failure happening, and I'll take a closer look at it. You can use `REQUIRE(next_table);` to make the test fail gracefully.

Reply:
9d99ce6